--- loop2.py	2022-10-25 15:16:27.545311021 +0200
+++ loop3.py	2022-10-25 15:18:31.777287961 +0200
@@ -1,10 +1,11 @@
 class Loop:
-    """Loop v2"""
+    """Loop v3"""
 
     _instance = None
 
     def __init__(self):
         self._running = False
+        self._scheduled = []
 
     @classmethod
     def get_current_loop(cls):
@@ -12,13 +13,25 @@
             cls._instance = Loop()
         return cls._instance
 
+    def run_step(self):
+        """Run a single step/tick of the loop"""
+        try:
+            task = self._scheduled.pop(0)  # fifo: extract first item
+            while task is not None:
+                task.step()
+                task = self._scheduled.pop(0)  # fifo: extract first item
+        except IndexError:
+            # list is empty
+            pass
+
     def run(self, coroutine):
         """Run a coroutine"""
         self._running = True
         step = 1
         while self._running:
-            print("Loop step", step)
+            print("Loop step", step, self._scheduled)
             try:
+                self.run_step()
                 next(coroutine)
                 step += 1
             except StopIteration as e:
@@ -28,3 +41,7 @@
     def stop(self):
         """Stop running the loop"""
         self._running = False
+
+    def schedule(self, task):
+        """Schedule a Task for the next step/tick"""
+        self._scheduled.append(task)
