--- task1.py	2022-11-02 10:06:06.566947275 +0100
+++ task2.py	2022-11-02 10:08:05.706252365 +0100
@@ -2,21 +2,26 @@
 
 from future import Future
 
+from loop import Loop
+
 
 class Task(Future):
-    """Task v1"""
+    """Task v2"""
 
     def __init__(self, coroutine: Generator[Any, None, Any], name: str):
         super().__init__(name)
         self._coroutine = coroutine
-        self.run_loop()
+        self._loop = Loop.get_current_loop()
+        self.schedule()
 
     def step(self) -> None:
         try:
             next(self._coroutine)
         except StopIteration as e:
             self.set_result(e.value)
+        else:
+            # no result yet
+            self.schedule()
 
-    def run_loop(self) -> None:
-        while not self.done():
-            self.step()
+    def schedule(self) -> None:
+        self._loop.schedule(self)
