--- loop3.py	2022-10-25 15:18:31.777287961 +0200
+++ loop4.py	2022-10-25 15:18:22.941289774 +0200
@@ -1,5 +1,8 @@
+from handle import Handle
+
+
 class Loop:
-    """Loop v3"""
+    """Loop v4"""
 
     _instance = None
 
@@ -16,10 +19,10 @@
     def run_step(self):
         """Run a single step/tick of the loop"""
         try:
-            task = self._scheduled.pop(0)  # fifo: extract first item
-            while task is not None:
-                task.step()
-                task = self._scheduled.pop(0)  # fifo: extract first item
+            handle = self._scheduled.pop(0)  # fifo: extract first item
+            while handle is not None:
+                handle.run()
+                handle = self._scheduled.pop(0)  # fifo: extract first item
         except IndexError:
             # list is empty
             pass
@@ -42,6 +45,6 @@
         """Stop running the loop"""
         self._running = False
 
-    def schedule(self, task):
-        """Schedule a Task for the next step/tick"""
-        self._scheduled.append(task)
+    def schedule(self, name, callback, *args):
+        """Schedule a callback for the next step/tick"""
+        self._scheduled.append(Handle(name, callback, args))
