--- task2.py	2022-10-26 09:41:50.994642454 +0200
+++ task3.py	2022-10-26 12:20:07.993060512 +0200
@@ -4,7 +4,7 @@
 
 
 class Task(Future):
-    """Task v2"""
+    """Task v3"""
 
     def __init__(self, coroutine, name):
         super().__init__(name)
@@ -22,4 +22,4 @@
             self.schedule()
 
     def schedule(self):
-        self._loop.schedule(self)
+        self._loop.schedule(self._name, self.step)
