--- loop4.py	2022-10-25 15:18:22.941289774 +0200
+++ loop5.py	2022-10-25 15:20:24.157262804 +0200
@@ -2,7 +2,7 @@
 
 
 class Loop:
-    """Loop v4"""
+    """Loop v5"""
 
     _instance = None
 
@@ -18,28 +18,43 @@
 
     def run_step(self):
         """Run a single step/tick of the loop"""
+        # execute all current known handles only.
+        # not the ones added while running the handle callbacks
+        scheduled = self._scheduled.copy()
+        self._scheduled.clear()
+
         try:
-            handle = self._scheduled.pop(0)  # fifo: extract first item
+            handle = scheduled.pop(0)  # fifo: extract first item
             while handle is not None:
                 handle.run()
-                handle = self._scheduled.pop(0)  # fifo: extract first item
+                handle = scheduled.pop(0)  # fifo: extract first item
         except IndexError:
             # list is empty
             pass
 
-    def run(self, coroutine):
-        """Run a coroutine"""
+        # a shorter and even more smarter version
+        #
+        # for _ in range(len(self._scheduled)):
+        #     handle = self._scheduled.pop(0)  # fifo: extract first item
+        #     handle.run()
+
+    def run_loop(self):
+        """Run the loop"""
         self._running = True
         step = 1
         while self._running:
             print("Loop step", step, self._scheduled)
-            try:
-                self.run_step()
-                next(coroutine)
-                step += 1
-            except StopIteration as e:
-                self._running = False
-                return e.value
+            self.run_step()
+            step += 1
+
+    def run(self, coroutine):
+        """Run a coroutine until it is done/completed"""
+        from task import Task  # avoid cyclic dependency
+
+        task = Task(coroutine, "Initial Task")
+        task.add_done_callback(self._done)
+        self.run_loop()
+        return task.result()
 
     def stop(self):
         """Stop running the loop"""
@@ -48,3 +63,6 @@
     def schedule(self, name, callback, *args):
         """Schedule a callback for the next step/tick"""
         self._scheduled.append(Handle(name, callback, args))
+
+    def _done(self, _future):
+        self.stop()
